feat: update technical priorities for 2023 (#47523)
nodejs/node: doc/contributing/technical-priorities.md

They will be updated regularly and will be reviewed by the next-10 team
and the TSC on a 6-month basis.

Version from the [mini-summit](https://github.com/nodejs/next-10/issues/1) on October 1st 2022.

## ESM

> _Present in: 2021_

Reviewer: ```suggestion

_Present in: 2021_

```

They will be updated regularly and will be reviewed by the next-10 team
and the TSC on a 6-month basis.

Version from the [mini-summit](https://github.com/nodejs/next-10/issues/1) on October 1st 2022.

> ## ESM

Reviewer 2: ```suggestion
## ES Modules (ESM)
```

develop and improve its ESM implementation to stay relevant and ensure
continued growth for the next 10 years.

## Suitable types for end-users

Using typings with JavaScript can allow a richer experience when using Visual
Studio Code (or any other IDEs) environments, more complete documentation
of APIs and the ability to identify and resolve errors earlier in the
development process. These benefits are important to a large number of Node.js
developers (maybe 50%).  Further typing support may be important
to enterprises that are considering expanding their preferred platforms to
include Node.js. It is, therefore, important that the Node.js project work
to ensure there are good typings available for the public Node.js APIs.

## Serverless

Serverless is a cloud computing model where the cloud provider manages the underlying infrastructure and automatically allocates resources as needed. Developers only need to focus on writing code for specific functions, which are executed as individual units of work in response to events. Node.js is one of the main technology used by developers in this field therefore it is crucial for us to provide a great solution.

## Small footprint

> Small software footprints refer to software that has a minimal impact on system resources such as memory and processing power. This can be achieved through various methods such as optimizing code, reducing the number of dependencies, or using lightweight frameworks. Smaller footprints can lead to faster startup times, reduced memory usage, and improved overall system performance. This is primordial for Node.js to be a lightweigh proposition inside the ecosystem as it is used across a wild variety of projects, from web application to IoT and serverless.

Reviewer 2: I’m not sure what “primordial” means in this context.

```suggestion
Small software footprints refer to software that has a minimal impact on system resources such as memory and processing power. This can be achieved through various methods such as optimizing code, reducing the number of dependencies, or using lightweight frameworks. Smaller footprints can lead to faster startup times, reduced memory usage, and improved overall system performance. This is fundamental for Node.js to be a lightweight proposition inside the ecosystem as it is used across a wild variety of projects, from web application to IoT and serverless.
```

Reviewer 3: +1 on using fundamental as the wording

> Small software footprints refer to software that has a minimal impact on system resources such as memory and processing power. This can be achieved through various methods such as optimizing code, reducing the number of dependencies, or using lightweight frameworks. Smaller footprints can lead to faster startup times, reduced memory usage, and improved overall system performance. This is primordial for Node.js to be a lightweigh proposition inside the ecosystem as it is used across a wild variety of projects, from web application to IoT and serverless.

Reviewer 4: ```suggestion
Small software footprints refer to software that has a minimal impact on system resources such as memory and processing power. This can be achieved through various methods such as optimizing code, reducing the number of dependencies, or using lightweight frameworks. Smaller footprints can lead to faster startup times, reduced memory usage, and improved overall system performance. This is primordial for Node.js to be a lightweight proposition inside the ecosystem as it is used across a wild variety of projects, from web application to IoT and serverless.
```

of the solution. This includes implementations of “host” APIs like WASI.

## Documentation

The current documentation is great for experienced developers or people
who are aware of what they are looking for. On the other hand, for
beginners this documentation can be quite hard to read and finding the
desired information is difficult. We must have documentation
that is suitable for beginners to continue the rapid growth in use.
This documentation should include more concrete examples and a learning
path for newcomers.

## Developers-first DX

Developer experience (DX) refers to the overall experience a developer has when working with a software development platform, framework, or tool. It encompasses all aspects of the developer's interactions with the system, from installation and configuration to writing code and debugging. A good DX prioritizes ease of use, efficiency, and productivity, and can lead to faster development times, higher quality code, and greater developer satisfaction. Factors that can impact DX include documentation, community support, testing tools, and integration with other systems.

------

> Version from the [mini-summit](https://github.com/nodejs/next-10/issues/76) on August 5th 2021.

Reviewer 2: Why do we need older versions below the current one? Can this file just represent the current priorities list, and the older versions are in the git history?

include Node.js. It is, therefore, important that the Node.js project work
to ensure there are good typings available for the public Node.js APIs.

## Serverless

> Serverless is a cloud computing model where the cloud provider manages the underlying infrastructure and automatically allocates resources as needed. Developers only need to focus on writing code for specific functions, which are executed as individual units of work in response to events. Node.js is one of the main technology used by developers in this field therefore it is crucial for us to provide a great solution.

Reviewer 5: Ultimately most serverless runtimes are closed-source forks of Node.js, patched to support the needs of their cloud. Given that Serverless vendors are not keen to participate in the development of Node.js, nor do they provide public feedback, I don't see how this priority could be translated into actual work.

I'm +1 to support whoever shows up and wants to collaborate with the Node.js project. Supporting companies that are not willing to give back is not something I can endorse.

I think a goal could be to have those companies start collaborating on the project, _but_ I don't think "serverless" could be a technical priority.

Serverless is a cloud computing model where the cloud provider manages the underlying infrastructure and automatically allocates resources as needed. Developers only need to focus on writing code for specific functions, which are executed as individual units of work in response to events. Node.js is one of the main technology used by developers in this field therefore it is crucial for us to provide a great solution.

The use of WebAssembly has been growing over the last few years.
To ensure Node.js continues to be part of solutions where a
subset of the solution needs the performance that WebAssembly can
deliver, Node.js must provide good support for running
WebAssembly components along with the JavaScript that makes up the rest
of the solution. This includes implementations of “host” APIs like WASI.
## Small footprint

## ESM

The CommonJS module system was one of the key components that led to the success
of Node.js in its first 10 years. ESM is the standard that has been adopted as
the equivalent in the broader JavaScript ecosystem and Node.js must continue to
develop and improve its ESM implementation to stay relevant and ensure
continued growth for the next 10 years.

> ## Support for features from the latest ECMAScript spec

Reviewer 3: I think we should leave this one in. Just because it was not in the current brainstorm does not mean it's still not important to the project.  I believe this one will continue to be important almost forever :)

> _Present in: 2021_

Reviewer: ```suggestion

_Present in: 2021_

```

## WebAssembly

_Present in: 2021_

The use of WebAssembly has been growing over the last few years.
To ensure Node.js continues to be part of solutions where a
subset of the solution needs the performance that WebAssembly can
deliver, Node.js must provide good support for running
WebAssembly components along with the JavaScript that makes up the rest
of the solution. This includes implementations of “host” APIs like WASI.

## ES Modules (ESM)

_Present in: 2021_

The CommonJS module system was one of the key components that led to the
success of Node.js in its first 10 years. ESM is the standard that has

> been adopted asthe equivalent in the broader JavaScript ecosystem and

Reviewer 3: ```suggestion
been adopted asthe equivalent in the broader JavaScript ecosystem and
```

security model so that consumers understand what threats are/are
not addressed by the Node.js runtime. We also need to provide
functions/features which help them limit attack surfaces even if it does
not result in 100% protection as this will still help organizations
manage their overall risk level.

## Better multithreaded support

Today's servers support multiple threads of concurrent execution.
Node.js deployments must be able to make full and efficient
use of the available resources. The right answer is often to use
technologies like containers to run multiple single threaded Node.js
instances on the same server. However, there are important use cases
where a single Node.js instance needs to make use of multiple threads
to achieve a performant and efficient implementation. In addition,
even when a Node.js instance only needs to consume a single thread to
complete its work there can be issues. If that work is long running,
blocking the event loop will interfere with other supporting work like
metrics gathering and health checks. Node.js
must provide good support for using multiple threads
to ensure the continued growth and success of Node.js.

> ## Single Executable Applications

Reviewer 3: same with this one.

> _Present in: 2021_

Reviewer: ```suggestion

_Present in: 2021_

```

> _Present in: 2021_

Reviewer: ```suggestion

_Present in: 2021_

```

> _Present in: 2021_

Reviewer: ```suggestion

_Present in: 2021_

```

## Developers-first DX

> Developer experience (DX) refers to the overall experience a developer has when working with a software development platform, framework, or tool. It encompasses all aspects of the developer's interactions with the system, from installation and configuration to writing code and debugging. A good DX prioritizes ease of use, efficiency, and productivity, and can lead to faster development times, higher quality code, and greater developer satisfaction. Factors that can impact DX include documentation, community support, testing tools, and integration with other systems.

Reviewer: ```suggestion
Developer experience (DX) refers to the overall experience a developer has when
working with a software development platform, framework, or tool. It encompasses
all aspects of the developer's interactions with the system, from installation
and configuration to writing code and debugging. A good DX prioritizes ease of
use, efficiency, and productivity, and can lead to faster development times,
higher quality code, and greater developer satisfaction. Factors that can impact
DX include documentation, community support, testing tools, and integration with
other systems.
```

document the use cases, problem determination methods and best
practices for those tools.

## Permissions/policies/security model

Organizations will only choose technologies that allow them to sufficiently
manage risk in their production deployments. For Node.js to
continue its growth in product/enterprise deployments we need to ensure
that we help them manage that risk. We must have a well-documented
security model so that consumers understand what threats are/are
not addressed by the Node.js runtime. We also need to provide
functions/features which help them limit attack surfaces even if it does
not result in 100% protection as this will still help organizations
manage their overall risk level.

> ## Better multithreaded support

Reviewer 3: This one seemed to be mentioned quite a bit in the Community survey so may be worth leaving in.

Reviewer 3: @sheplu I don't think this comment was addressed. Can you take another look.

Author: good catch !

for using Node.js in these situations. This is important to support
the expansion of where/when Node.js is used in building solutions.

## Serverless

Serverless is a cloud computing model where the cloud provider manages the
underlyinginfrastructure and automatically allocates resources as
needed. Developers only need to focus on writing code for specific
functions, which are executed as individual units of work in response to
events. Node.js is one of the main technology used by developers in
this field therefore it is crucial for us to provide a great solution.

## Small footprint

Small software footprints refer to software that has a minimal impact on
system resources such as memory and processing power. This can be achieved
through various methods such as optimizing code, reducing the number of
dependencies, or using lightweight frameworks. Smaller footprints can lead
to faster startup times, reduced memory usage, and improved overall system
performance. This is fundamental for Node.js to be a lightweight proposition
inside the ecosystem as it is used across a wild variety of projects, from
web application to IoT and serverless.

> ## Developers-first DX

Reviewer 5: I think this should mention TS Support, or TS support should be its own bucket.

Author: added a quick sentence. I think if ts will be one main point on the next 10 survey this year we can move that to its own dedicated part

Developer experience (DX) refers to the overall experience a developer has when
working with a software development platform, framework, or tool. It encompasses
all aspects of the developer's interactions with the system, from installation
and configuration to writing code and debugging. A good DX prioritizes ease

> of use,efficiency, and productivity, and can lead to faster development times,

Reviewer: ```suggestion
of use, efficiency, and productivity, and can lead to faster development times,
```

## Small footprint

Small software footprints refer to software that has a minimal impact on
system resources such as memory and processing power. This can be achieved
through various methods such as optimizing code, reducing the number of
dependencies, or using lightweight frameworks. Smaller footprints can lead
to faster startup times, reduced memory usage, and improved overall system
performance. This is fundamental for Node.js to be a lightweight proposition
inside the ecosystem as it is used across a wild variety of projects, from
web application to IoT and serverless.

## Developers-first DX

Developer experience (DX) refers to the overall experience a developer has when
working with a software development platform, framework, or tool. It encompasses
all aspects of the developer's interactions with the system, from installation
and configuration to writing code and debugging. A good DX prioritizes ease
of use, efficiency, and productivity, and can lead to faster development times,
higher quality code, and greater developer satisfaction. Factors that can
impact DX include documentation, community support, testing tools,
and integration with other systems.

> As TypeScript becomes usage continue to grow and the technology is more and 
> more prominent, improving its support is vital to provides a great developer 
> experience for newcomers or experienced users.

Reviewer: ```suggestion
As TypeScript usage continues to grow and gains more prominence in the
ecosystem, enhancing its support is essential for delivering an improved
developer experience for newcomers and experienced users alike.
```

Author: done @aduh95
